### DISCUSSION TITLE
Double validation errors displayed when view model implements INotifyDataErrorInfo

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#10480] [Q&A] [answered]
I have a form with a required field. I use the `[Required]` attribute on the view model property and got a reasonable UI:
![image](https://user-images.githubusercontent.com/3548/221371522-04284884-8dbd-4ebe-98fd-4e577fc7f590.png)

Here is the view and view model:

```xaml
	<ComboBox Items="{Binding Varieties}" SelectedItem="{Binding Variety}" Name="VarietyPicker" />
```

```cs
public class MainWindowViewModel : ViewModelBase {
    private string? variety;

    [Required]
    public string? Variety {
        get => this.variety;
        set => this.RaiseAndSetIfChanged(ref this.variety, value);
    }

    public List<string> Varieties { get; } = new() { "A", "B" };
}
```

*But*, when my `ViewModelBase` class implements `INotifyDataErrorInfo`, I get double errors:

![image](https://user-images.githubusercontent.com/3548/221371623-3826886b-dcd3-4863-b303-548daaca8af7.png)

```cs
public class ViewModelBase : ReactiveObject, INotifyDataErrorInfo {
    public event EventHandler<DataErrorsChangedEventArgs>? ErrorsChanged;

    public bool HasErrors => this.GetErrors(null).Any() || this.GetValidatedProperties().SelectMany(p => this.GetErrors(p.Name).Cast<ValidationResult>()).Any();

    IEnumerable INotifyDataErrorInfo.GetErrors(string? propertyName) => this.GetErrors(propertyName);

    public virtual IEnumerable<ValidationResult> GetErrors(string? propertyName)
    {
        if (string.IsNullOrEmpty(propertyName))
        {
            return Enumerable.Empty<ValidationResult>();
        }

        List<ValidationResult> validationResults = new();

        PropertyInfo? property = this.GetType().GetProperties().FirstOrDefault(p => p.Name == propertyName) ?? throw new ArgumentException("No property by that name.");

        ValidationContext validationContext = new(this)
        {
            MemberName = propertyName,
        };

        Validator.TryValidateProperty(property.GetValue(this), validationContext, validationResults);
        return validationResults;
    }

    private IEnumerable<PropertyInfo> GetValidatedProperties() => this.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance).Where(pi => pi.GetCustomAttributes<ValidationAttribute>(true).Any());
}
```

I think my implementation is reasonable. My view model is responsible for reporting errors for individual properties, right? And with the debugger I can confirm that only 1 validation result is returned, yet it gets doubled. If my method returns 0 results, no validation error is displayed at all. So I'm forced to choose between 0 errors and 2 errors being displayed to the user when implementing this interface. 
I implement the interface because I have some complex validations to run in some cases. But I want to leverage the validation attributes whenever I can. 

Suggestions on how to get this down to just one validation error displayed to the user?

**Comment by timunie** [comment, marked as answer]
Check the docs. You need to remove one of the validators.

https://docs.avaloniaui.net/docs/data-binding/data-validation#manage-validationplugins

**Reply by AArnott** [reply]
Thanks!

**Reply by AArnott** [reply]
Just tried it though, and it doesn't compile. The docs say to add this:

```cs
 ExpressionObserver.DataValidators.RemoveAll(x => x is DataAnnotationsValidationPlugin);
```

But `ExpressionObserver` doesn't bind. Was it renamed in a recent Avalonia version? I'm on daily builds.

**Reply by Mrxx99** [reply]
Yes it was changed, you can use this since 11preview5:  `BindingPlugins.DataValidators.RemoveAt(0);`
